scripts/test/Muon/muon_group_pair_context_test.py

```diff
@@ -0,0 +1,134 @@
+#     & Institut Laue - Langevin
+# SPDX - License - Identifier: GPL - 3.0 +
+import sys
+import unittest
+from Muon.GUI.Common.muon_group_pair_context import MuonGroupPairContext
+from Muon.GUI.Common.observer_pattern import Observer
+from Muon.GUI.Common.muon_group import MuonGroup
+from Muon.GUI.Common.muon_pair import MuonPair
+
+if sys.version_info.major < 2:
+    from unittest import mock
+else:
+    import mock
+
+class MuonGroupPairContextTest(unittest.TestCase):
+    def setUp(self):
+        self.context = MuonGroupPairContext()
+
+    def test_can_be_created(self):
+        self.assertTrue(self.context)
+
+    def test_groups_and_pairs_initially_empty(self):
+        self.assertEqual(self.context.groups, [])
+        self.assertEqual(self.context.pairs, [])
+
+    def test_group_can_be_added(self):
+        group = MuonGroup('group_1', [1,3,5,7,9])
+
+        self.context.add_group(group)
+
+        self.assertEqual(self.context['group_1'], group)
+
+    def test_non_group_cannot_be_added(self):
+        pair = MuonPair('pair_1')
+
+        self.assertRaises(AssertionError, self.context.add_group, pair)
+
+    def test_cannot_add_group_with_duplicate_name(self):
+        group_1 = MuonGroup('group_1', [1, 3, 5, 7, 9])
+        group_2 = MuonGroup('group_1', [1, 3, 5, 7, 9])
+
+        self.context.add_group(group_1)
+
+        self.assertRaises(ValueError, self.context.add_group, group_2)
+
+    def test_pair_can_be_added(self):
+        pair = MuonPair('pair_1')
+
+        self.context.add_pair(pair)
+
+        self.assertEqual(self.context['pair_1'], pair)
+
+    def test_non_pair_cannot_be_added(self):
+        pair = MuonPair('pair_1')
+
+        self.assertRaises(AssertionError, self.context.add_group, pair)
+
+    def test_cannot_add_pair_with_duplicate_name(self):
+        pair_1 = MuonPair('pair')
+        pair_2 = MuonPair('pair')
+
+        self.context.add_pair(pair_1)
+
+        self.assertRaises(ValueError, self.context.add_pair, pair_2)
+
+    def test_show_adds_group_or_pair_to_ADS(self):
+        group_1 = MuonGroup('group_1', [1, 3, 5, 7, 9])
+        group_2 = MuonGroup('group_2', [1, 3, 4, 7, 9])
+        pair = MuonPair('pair_1')
+        group_1.show = mock.MagicMock()
+        pair.show = mock.MagicMock()
+        self.context.add_group(group_1)
+        self.context.add_group(group_2)
+        self.context.add_pair(pair)
+
+        self.context.show('group_1', [12345])
+        self.context.show('pair_1', [12345])
+
+        group_1.show.assert_called_once_with(str([12345]))
+        pair.show.assert_called_once_with(str([12345]))
+
+    def test_group_names_returns_ordered_list_of_names(self):
+        group_1 = MuonGroup('group_1', [1, 3, 5, 7, 9])
+        group_2 = MuonGroup('group_2', [1, 3, 4, 7, 9])
+        group_3 = MuonGroup('group_3', [1, 3, 4, 7, 9])
+
+        self.context.add_group(group_1)
+        self.context.add_group(group_2)
+        self.context.add_group(group_3)
```

Comment:
if you really want to test that it keeps the order you stored them in, then store them in a non numeric or alphabetic way. e.g. 2,3,1

Reply:
Good idea